File: interfaces/matlab/examples/laplacian1d.m
% Copyright (c) 2024 Lawrence Livermore National Security, LLC and other
% HYPRE Project Developers. See the top-level COPYRIGHT file for details.
%
% SPDX-License-Identifier: MIT
%
%LAPLACIAN1D Solve a small 1D Poisson problem with hypredrive.
%
% Run from a build tree containing the hypredrive MEX file:
%
%   addpath('/path/to/build');
%   laplacian1d

n = 64;
e = ones(n, 1);
A = spdiags([-e, 2 * e, -e], -1:1, n, n);
b = ones(n, 1);

opts = hypredrive_options( ...
    'solver', 'pcg', ...
    'preconditioner', 'amg', ...
    'pcg', struct('max_iter', 100, 'relative_tol', 1.0e-8), ...
    'amg', struct('max_iter', 1, 'tolerance', 0.0, 'print_level', 0));

[x, info] = hypredrive_solve(A, b, opts);

fprintf('iterations: %d\n', info.iterations);
fprintf('setup time: %.6e s\n', info.setup_time);
fprintf('solve time: %.6e s\n', info.solve_time);
fprintf('solution l2 norm: %.6e\n', info.solution_norm);
fprintf('relative residual: %.6e\n', norm(b - A * x) / norm(b));

File: interfaces/julia/examples/laplacian1d.jl
# Copyright (c) 2024, Lawrence Livermore National Security, LLC.
# See the top-level LICENSE and NOTICE files for details.

using HypreDrive
using LinearAlgebra
using SparseArrays

function laplacian1d(n::Integer)
    main = fill(2.0, n)
    off = fill(-1.0, n - 1)
    return spdiagm(-1 => off, 0 => main, 1 => off)
end

# Hypredrive YAML overrides given after -a/--args, e.g.
# julia laplacian1d.jl 64 -a --solver:pcg:max_iter 100
overrides_at = findfirst(a -> a in ("-a", "--args"), ARGS)
positional = overrides_at === nothing ? ARGS : ARGS[1:overrides_at - 1]
n = length(positional) >= 1 ? parse(Int, positional[1]) : 128
A = laplacian1d(n)
b = ones(n)

input_args = overrides_at === nothing ? nothing :
             vcat(String[hypredrive_options()], ARGS[overrides_at:end])
x, info = hypredrive_solve(A, b; input_args=input_args)
relres = norm(A * x - b) / norm(b)

println("unknowns: ", n)
println("iterations: ", info.iterations)
println("relative residual: ", relres)
